pkg/api/validate/dynamic/serviceprincipal.go

```diff
@@ -25,7 +25,7 @@ func (dv *dynamic) ValidateServicePrincipal(ctx context.Context, clientID, clien
 	c := &azureclaim.AzureClaim{}
 	_, _, err = p.ParseUnverified(token.OAuthToken(), c)
 	if err != nil {
-		return err
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", "The provided service principal generated an invalid token.")
```

Comment:
Unfortunately, the docs for ParseUnverified don't specify what errors are returned. Do we want to at least log them before masking/returning them like this?

Reply:
Agreed, I changed this and the similar files to now produce a log with the error before returning the cloudError type

Reply:
`dynamic.ValidateServicePrincipal` is not only called from the [API SP validation](https://github.com/Azure/ARO-RP/blob/7447d1b01df6974a8069b265f85fa59d7b15a0c5/pkg/api/validate/openshiftcluster_validatedynamic.go#L90) but also from the [ARO operator when performing SP validation](https://github.com/Azure/ARO-RP/blob/7447d1b01df6974a8069b265f85fa59d7b15a0c5/pkg/operator/controllers/checkers/serviceprincipalchecker/checker.go#L60), so this change would mean that [the error message in OpenShift would be set to the CloudError message](https://github.com/Azure/ARO-RP/blob/7447d1b01df6974a8069b265f85fa59d7b15a0c5/pkg/operator/controllers/checkers/serviceprincipalchecker/controller.go#L104) instead of the actual root cause error.

I think this needs to be tested, so that we can see the difference between the two and decide whether we need to add conditional logic in the ARO Operator to parse the CloudError. Potentially only displaying the message segment after the last `:`, i.e. "The provided service principal generated an invalid token."